packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -61,19 +70,53 @@ export class ConnectionStateHandler extends EventEmitterWithErrorHandling<IConne
         super();
     }
 
-    public receivedAddMemberEvent(clientId: string, quorum: Quorum) {
+    // This is true when this client submitted any ops.
+    public setDirtyState() {
+        this.isDirty = true;
+    }
+
+    public receivedAddMemberEvent(clientId: string) {
         // This is the only one that requires the pending client ID
         if (clientId === this.pendingClientId) {
+            // Wait for previous client to leave the quorum before firing "connected" event.
+            if (this.prevClientLeftTimer !== undefined && this.prevClientLeftTimer.hasTimer) {
+                this.waitEvent = PerformanceEvent.start(this.logger, {
+                    eventName: "WaitBeforeClientLeave",
+                    waitOnClientId: this._clientId,
+                    hadOutstandingOps: this.handler.shouldClientJoinWrite(),
+                });
+                this.waitingClientId = clientId;
+            } else {
+                this.setConnectionState(ConnectionState.Connected);
+            }
+        }
+    }
+
+    private clientLeaveWaitEnded(leaveReceived: boolean) {
+        // Move to connected state only if there was a client waiting for client leave. It may happen that
+        // during wait, the waiting client again got Disconnected/Connecting and then we don't want to
+        // move to connected state here for a non waiting client.
+        if (this.waitingClientId !== undefined) {
+            assert(this.waitingClientId === this.pendingClientId, "Pending Client Id should match the waiting client");
```

Comment:
I do not think this assert is correct. As we might reconnect multiple times before we see leave from first connection.
I believe pendingClientId always tracks current connection, and it might even be undefined if we have no connection at the moment

Reply:
I think same applies to logic around this.waitEvent - it's only there if we have first reconnection active. What if we have no ordering service connection, but received leave through GetDeltas workflow (i.e. storage)?

Reply:
So waiting client id is different from pending client id. Here is the def in the comments:
    // This is client id of client for which we have received the addMember event but we are waiting on some previous
    // client to leave before moving to Connected state.
while pending client id is one for which we received the connect event.
So the idea is that based on leave whether through timeout or actual leave message(through ordering service or storage), if we had any current client waiting on it(For which we received connected event), then only its usage come into picture as only then we would check that if waitingClientId  is set, then it should be equal to pending client id. Otherwise just clear the timer or it automatically clears itself on timeout and we are done waiting. Even if the ops are processed through storage, if we have timer running and leave is processed, we would clear the timer and then later when the connection will happen we will straight away move to Connected state instead of waiting.

Reply:
Got it!